File: backend/chunker.py
import hashlib
from pathlib import Path
from typing import Iterator

from .config import settings
from .models import CodeChunk

EXTENSION_TO_LANGUAGE: dict[str, str] = {
    ".py": "python",
    ".ts": "typescript",
    ".tsx": "typescript",
    ".js": "javascript",
    ".jsx": "javascript",
    ".vue": "vue",
    ".kt": "kotlin",
    ".java": "java",
    ".cs": "csharp",
    ".go": "go",
    ".rs": "rust",
    ".rb": "ruby",
    ".php": "php",
    ".swift": "swift",
    ".cpp": "cpp",
    ".c": "c",
    ".h": "c",
    ".hpp": "cpp",
    ".html": "html",
    ".css": "css",
    ".scss": "scss",
    ".json": "json",
    ".yaml": "yaml",
    ".yml": "yaml",
    ".toml": "toml",
    ".md": "markdown",
    ".sh": "bash",
}

SKIP_DIRS = {
    ".git",
    ".svn",
    "node_modules",
    "__pycache__",
    ".venv",
    "venv",
    "dist",
    "build",
    ".next",
    ".nuxt",
    "coverage",
    ".pytest_cache",
    ".mypy_cache",
    ".ruff_cache",
    "target",
    # C# build output
    "bin",
    "obj",
    "packages",
    ".vs",
    "TestResults",
    "publish",
    "artifacts",
    # Vendored / embedded libraries
    "~ExternalLibraries",
    "Dommel",
    # Capacitor / mobile native
    "android",
    "ios",
    "electron",
    # CI/CD
    "fastlane",
}

SKIP_DIR_PREFIXES = (
    "Intra.Old.",
)  # legacy C# code, not relevant to current work

SKIP_EXTENSIONS = {
    ".lock",
    ".png",
    ".jpg",
    ".jpeg",
    ".gif",
    ".svg",
    ".ico",
    ".woff",
    ".woff2",
    ".ttf",
    ".eot",
    ".pdf",
    ".zip",
    ".tar",
    ".gz",
    ".map",
    ".min.js",
    ".min.css",
}

SKIP_FILENAMES = {
    # Lockfiles (various extensions, all auto-generated)
    "pnpm-lock.yaml",
    "yarn.lock",
    "package-lock.json",
    "Gemfile.lock",
    "Pipfile.lock",
    "composer.lock",
    "poetry.lock",
    "go.sum",
}


def _should_skip_dir(name: str) -> bool:
    if name in SKIP_DIRS or name.startswith("."):
        return True
    return any(name.startswith(p) for p in SKIP_DIR_PREFIXES)


def _get_language(path: Path) -> str | None:
    suffix = path.suffix.lower()
    # Handle compound extensions like .min.js
    if path.name.endswith(tuple(SKIP_EXTENSIONS)):
        return None
    return EXTENSION_TO_LANGUAGE.get(suffix)


def _load_gitignore(root: Path):
    gitignore_path = root / ".gitignore"
    if gitignore_path.exists():
        try:
            import gitignore_parser  # type: ignore[import-untyped]

            return gitignore_parser.parse_gitignore(gitignore_path)
        except Exception:
            pass
    return None


def _chunk_lines(
    lines: list[str], start_offset: int = 0
) -> list[tuple[int, int, str]]:
    """Split lines into overlapping chunks.
    Returns (start_line, end_line, content) tuples.
    """
    size = settings.chunk_size
    overlap = settings.chunk_overlap
    step = size - overlap
    chunks = []
    i = 0
    while i < len(lines):
        chunk_lines = lines[i : i + size]
        content = "".join(chunk_lines).strip()
        if content:
            chunks.append(
                (
                    start_offset + i + 1,
                    start_offset + i + len(chunk_lines),
                    content,
                )
            )
        i += step
    return chunks


def walk_repo_files(root_path: str) -> Iterator[tuple[str, Path]]:
    """
    Yield (rel_path, abs_path) for every indexable file in the repo.
    Applies the same skip rules as chunk_repo — no file reading or chunking.
    """
    root = Path(root_path)
    matches_gitignore = _load_gitignore(root)

    for path in sorted(root.rglob("*")):
        if not path.is_file():
            continue
        if any(
            _should_skip_dir(part)
            for part in path.relative_to(root).parts[:-1]
        ):
            continue
        if path.name in SKIP_FILENAMES:
            continue
        if matches_gitignore and matches_gitignore(str(path)):
            continue
        if _get_language(path) is None:
            continue
        rel_path = str(path.relative_to(root)).replace("\\", "/")
        yield rel_path, path


def chunk_repo(
    project_id: str,
    root_path: str,
    only_files: set[str] | None = None,
) -> Iterator[CodeChunk]:
    """
    Yield CodeChunk objects for every indexable file in the repo.
    Pass only_files to restrict chunking to a specific set of rel_paths.
    """
    for rel_path, path in walk_repo_files(root_path):
        if only_files is not None and rel_path not in only_files:
            continue

        try:
            text = path.read_text(encoding="utf-8", errors="ignore")
        except Exception:
            continue

        lines = text.splitlines(keepends=True)

        for start_line, end_line, content in _chunk_lines(lines):
            chunk_id = hashlib.sha256(
                f"{project_id}:{rel_path}:{start_line}:{end_line}".encode()
            ).hexdigest()

            yield CodeChunk(
                id=chunk_id,
                project_id=project_id,
                file_path=rel_path,
                content=content,
                start_line=start_line,
                end_line=end_line,
                language=_get_language(path) or "",
            )

File: src/chunker.ts
import { createHash } from "crypto";
import { readFileSync, readdirSync } from "fs";
import { join, relative, extname, basename } from "path";
import { config } from "./config.js";
import { loadCanonicalIgnoreRules } from "./ignore-rules.js";
import { normalizeContent } from "./normalize.js";
import type { CodeChunk, RawChunk, StampedChunk } from "./types.js";


const EXTENSION_TO_LANGUAGE: Record<string, string> = {
  ".py": "python",
  ".ts": "typescript",
  ".tsx": "tsx",
  ".js": "javascript",
  ".jsx": "javascript",
  ".vue": "vue",
  ".kt": "kotlin",
  ".java": "java",
  ".cs": "csharp",
  ".go": "go",
  ".rs": "rust",
  ".rb": "ruby",
  ".php": "php",
  ".swift": "swift",
  ".cpp": "cpp",
  ".c": "c",
  ".h": "c",
  ".hpp": "cpp",
  ".html": "html",
  ".css": "css",
  ".scss": "scss",
  ".json": "json",
  ".yaml": "yaml",
  ".yml": "yaml",
  ".toml": "toml",
  ".md": "markdown",
  ".sh": "bash",
};

const SKIP_DIRS = new Set([
  ".git",
  ".svn",
  "node_modules",
  "__pycache__",
  ".venv",
  "venv",
  "dist",
  "build",
  ".next",
  ".nuxt",
  "coverage",
  ".pytest_cache",
  ".mypy_cache",
  ".ruff_cache",
  "target",
  "bin",
  "obj",
  "packages",
  ".vs",
  "TestResults",
  "publish",
  "artifacts",
  "~ExternalLibraries",
  "Dommel",
  "vendor",
  "android",
  "ios",
  "electron",
  "fastlane",
]);

const SKIP_DIR_PREFIXES = ["Intra.Old."];

const SKIP_EXTENSIONS = new Set([
  ".lock",
  ".png",
  ".jpg",
  ".jpeg",
  ".gif",
  ".svg",
  ".ico",
  ".woff",
  ".woff2",
  ".ttf",
  ".eot",
  ".pdf",
  ".zip",
  ".tar",
  ".gz",
  ".map",
]);

const SKIP_FILENAMES = new Set([
  "pnpm-lock.yaml",
  "yarn.lock",
  "package-lock.json",
  "Gemfile.lock",
  "Pipfile.lock",
  "composer.lock",
  "poetry.lock",
  "go.sum",
]);

function shouldSkipDir(name: string): boolean {
  if (name.startsWith(".")) return true;
  if (SKIP_DIRS.has(name)) return true;
  return SKIP_DIR_PREFIXES.some((p) => name.startsWith(p));
}

export function getLanguage(filename: string): string | null {
  const name = basename(filename);
  if (name.endsWith(".min.js") || name.endsWith(".min.css")) return null;
  // Auto-generated C# files
  if (name.endsWith(".g.cs") || name.endsWith(".designer.cs") || name.endsWith(".Designer.cs") || name.endsWith(".generated.cs")) return null;
  const ext = extname(name).toLowerCase();
  if (SKIP_EXTENSIONS.has(ext)) return null;
  return EXTENSION_TO_LANGUAGE[ext] ?? null;
}

export function chunkLines(
  lines: string[],
  startOffset = 0
): Array<{ start: number; end: number; content: string }> {
  const size = config.chunkSize;
  const overlap = config.chunkOverlap;
  const step = size - overlap;
  const result: Array<{ start: number; end: number; content: string }> = [];
  let i = 0;
  while (i < lines.length) {
    const slice = lines.slice(i, i + size);
    const content = slice.join("").trim();
    if (content) {
      result.push({
        start: startOffset + i + 1,
        end: startOffset + i + slice.length,
        content,
      });
    }
    i += step;
  }
  return result;
}

export function* walkRepoFiles(
  rootPath: string,
  projectId?: string,
  sourceId?: string
): Generator<{ relPath: string; absPath: string }> {
  const rules = loadCanonicalIgnoreRules(rootPath, projectId, sourceId);
  const { forcePatterns } = rules;

  function dirMightContainForceIncludes(dirRelPath: string): boolean {
    const prefix = dirRelPath + "/";
    return forcePatterns.some(p => p === dirRelPath || p === prefix || p.startsWith(prefix));
  }

  function* walk(dir: string): Generator<{ relPath: string; absPath: string }> {
    try {
      for (const entry of readdirSync(dir, { withFileTypes: true })) {
        const name = String(entry.name);
        const absPath = join(dir, name);
        const relPath = relative(rootPath, absPath).replace(/\\/g, "/");

        if (entry.isDirectory()) {
          if (shouldSkipDir(name) && !dirMightContainForceIncludes(relPath)) continue;
          yield* walk(absPath);
        } else if (entry.isFile()) {
          const forceIn = rules.isForceIncluded(relPath);
          if (!forceIn && SKIP_FILENAMES.has(name)) continue;
          if (!forceIn && getLanguage(name) === null) continue;
          if (rules.shouldIgnore(relPath)) continue;
          yield { relPath, absPath };
        }
      }
    } catch {
      // unreadable directory — skip
    }
  }

  yield* walk(rootPath);
}

function makeChunkId(
  projectId: string,
  sourceId: string,
  itemPath: string,
  itemUrl: string,
  itemType: string,
  content: string
): string {
  return createHash("sha256")
    .update(projectId + "\0" + sourceId + "\0" + itemPath + "\0" + itemUrl + "\0" + itemType + "\0" + content)
    .digest("hex");
}

/** Central chunk-ID stamping: accepts a RawChunk, returns a StampedChunk with chunk_id set. */
export function stampChunkId(raw: RawChunk): StampedChunk {
  const chunk_id = makeChunkId(
    raw.project_id,
    raw.source_id,
    raw.item_path,
    raw.item_url,
    raw.item_type,
    raw.content,
  );
  return { ...raw, chunk_id } as StampedChunk;
}

export function* chunkRepo(
  projectId: string,
  sourceId: string,
  rootPath: string,
  onlyFiles?: Set<string>
): Generator<CodeChunk> {
  for (const { relPath, absPath } of walkRepoFiles(rootPath, projectId, sourceId)) {
    if (onlyFiles !== undefined && !onlyFiles.has(relPath)) continue;

    let text: string;
    try {
      text = normalizeContent(readFileSync(absPath, "utf8"));
    } catch {
      continue;
    }

    const language = getLanguage(basename(absPath)) ?? "";
    const lines = text.split(/^/m);

    for (const { start, end, content } of chunkLines(lines)) {
      yield stampChunkId({
        project_id: projectId,
        source_id: sourceId,
        item_path: relPath,
        item_url: "",
        item_type: "code",
        content,
        start_line: start,
        end_line: end,
        language,
        symbol_name: "",
      }) as CodeChunk;
    }
  }
}
